examples/fourier_transform_playground.py

```diff
@@ -0,0 +1,151 @@
+
+    The wave phases are offset by the current time multiplied by an
+    arbitrary speed value; this generates an animated wave if called repeatedly.
+    """
+    if not wave_args:
+        # this happens on yielding from the thread, no need to update anything
+        return
+
+    t = time()
+    waves = {
+        wave_id: wave_2d(frequency, angle, phase_shift + t * speed * 100) if frequency else None
+        for wave_id, (frequency, angle, phase_shift, speed) in wave_args.items()
+    }
+
+    to_add = [d for d in waves.values() if d is not None]
+    if to_add:
+        mean = np.mean(to_add, axis=0)
+        ft = fftshift(fft2(mean))
+        power_spectrum = abs(ft)
+        phase = np.angle(ft) * power_spectrum
+        power_spectrum = np.log10(power_spectrum + 10)
+    else:
+        mean = power_spectrum = phase = None
+
+    # for visualisation, it's clearer to use:
+    # phase * ps instead of phase
+    # and log10(ps + 1) instead of ps
+    update_layer('phase * power_spectrum', phase, colormap=('blue', 'black', 'red'))
+    update_layer('log10(power_spectrum + 1)', power_spectrum)
+    update_layer('mean', mean)
+
+    for name, data in waves.items():
+        update_layer(f'wave {name}', data)
+
+
+@thread_worker(connect={"yielded": combine_and_set_data})
+def update_viewer():
+    # keep track of each wave in a dictionary by id, this way we can modify/remove
+    # existing waves or add new ones
+    wave_args = {}
+    new_params = None
+    while True:
+        sleep(1 / FPS)
+        # see https://napari.org/stable/guides/threading.html#full-two-way-communication
+        # this receives new_params from thread.send() and yields {} for the `yielded` callback
+        new_params = yield wave_args
+        if new_params is not None:
+            # note that these come from thread.send() in moving_wave()!
+            wave_id, *args = new_params
+            wave_args[wave_id] = args
+        yield wave_args
+
+
+# start the thread responsible for updating the viewer
+thread = update_viewer()
+
+
+@magic_factory(
+    auto_call=True,
+    frequency={'widget_type': 'FloatSlider', 'min': 0, 'max': 1, 'step': 0.01},
+    angle={'widget_type': 'Slider', 'min': 0, 'max': 180},
+    phase_shift={'widget_type': 'Slider', 'min': 0, 'max': 180},
+    speed={'widget_type': 'FloatSlider', 'min': -10, 'max': 10, 'step': 0.1},
+)
+def moving_wave(
+    wave_id: int = 0,
+    frequency: float = 0.2,
+    angle: int = 0,
+    phase_shift: int = 0,
+    speed: float = 1,
+    run=True,
+):
+    """Send new parameters to the listening thread to update the 2D waves.
+
+    The `run` checkbox can be disabled to stop sending values to the thread
+    while changing parameters.
+    """
+    if run:
+        thread.send((wave_id, frequency, angle, phase_shift, speed))
+
+
+wdg = moving_wave()
+
+# add the widget to the window and run it once
+viewer.window.add_dock_widget(wdg, area='bottom')
+wdg()
+
+napari.run()
```

Comment:
I think this should be protected by `if __name__ == '__main__':`, too? I'm a bit confused why it even works LOL

Reply:
The issue is that to define some of the callbacks, the thread needs to already be spawned... so I would have to stick a lot of stuff inside the `if __name__` block... not sure that helps much :P